### DISCUSSION TITLE
Rust language ideas

### DISCUSSION THREAD

**Discussion by hydroper** [airsdk/Adobe-Runtime-Support#2774] [Ideas]
If you consider inspiring in the [`wasm2swf`](https://github.com/brion/wasm2swf) GitHub in the future anytime, we could leave some ideas here. This project potentially allows languages other than Rust with Adobe AIR, but it might be better to rewrite this tool in Rust using [`wasmparser`](https://lib.rs/wasmparser) (by [Bytecode Alliance](https://bytecodealliance.org)) and [Ruffle's `swf`](https://lib.rs/swf).

This is similiar to the _ActionScript Language Ideas_ discussion, but for Rust, if potentially supported via `wasm2swf`. For newcomers, Rust supports procedural macros (like Haxe), algebraic data types (like Haxe) and much more. To me it feels like Haxe is an evolution of ActionScript taking unnecessary features out (like E4X and qualifiers). Rust is completely different though, so it's interesting taking a look.

**Comment by hydroper** [comment]
# Display list

Modern languages like Rust and Go favor composition over inheritance. In Rust, a `struct` cannot inherit another one.

Emulating structure inheritance is possible, but it's counterintuitive. Instead a pattern similiar to a gaming's entity-component-system (ECS) pattern can be used or more general. But you might guess that it won't cover what AIR has, like inherited properties. Well, the Bevy Engine for the Rust language uses _children_ entities and some kinds of entities are used as _inheritable_ properties.

There might be some terminology that fits well with AIR's display list.

#### XML

A XML-way of writing a display list can be supported by defining a procedural macro.

#### More on the pattern

Bevy Engine uses some different ways of manipulating entities, so you may want to take a look there.

**Comment by ajwfrost** [comment]
This is an interesting one - I hadn't really thought about anyone wanting to convert webassembly into ActionScript byte code! My feeling has more been that we could use LLVM as an intermediary and then have different front ends and back ends so e.g. you could compile C++ to LLVM and then to ABC - which is what Alchemy/Crossbridge does - or you could compile AS3 to LLVM and then to machine code - which is what happens when targeting iOS. If we could separate out each of those pieces then it could be handy.. 

The challenge seems to be that these intermediate languages do need to have knowledge of both the source and target systems, in order to know what options there are for some of the commands and structures - similarly, the JIT within AIR uses a low-level intermediate representation when turning ABC into machine code, and this is configured at build-time for the target environment. See https://github.com/adobe/avmplus/blob/master/nanojit/LIRopcode.tbl and https://github.com/adobe/avmplus/blob/master/core/CodegenLIR.cpp for some of how this all works...

Going back to your original thought of using a wasm2swf or wasmparser approach to turn something that's already turned into WebAssembly into SWF... there may be some extra complications in terms of the interactions with JavaScript and the whole bootstrapping of WebAssembly that's required from the browser side. But it's probably viable ..  I'd hope that these sorts of capabilities could be achieved by adapting the existing open source tools, sounds very intriguing through!

thanks

**Reply by hydroper** [reply]
I'm pretty inexperienced with LLVM, looks pretty advanced to me! If we were to go the LLVM route, at least I guess it involves...

- Creating a new target triple (`abc-adobe-unknown`) for `rustup` (`rustup` is the Rust installer or manager)
- Updating the Rust `std` library for handling of our AIR target. Most regions of the `std` (and `core`) are platform-agnostic; the only things I know that aren't are things like `std::time::SystemTime::now()`, which panics for WebAssembly.

I've never tried it, but here are all relevant docs:

- [Proposing an already-contributed target](https://doc.rust-lang.org/rustc/target-tier-policy.html#adding-a-new-target)
- [Custom targets](https://doc.rust-lang.org/rustc/targets/custom.html)
  - [More in-depth information](https://docs.rust-embedded.org/embedonomicon/custom-target.html)
 
If we were to go the `.wasm` route, I know that:

- In `std`, things like `std::time::SystemTime::now()` panic for WebAssembly. In the case of the WebAssembly target (`wasm32-unknown-unknown`), `std` doesn't implement them, so we're safe. Some packages in the `crates.io` registry may use `wasm_bindgen` for using things like `Date.now()` or `Promise`. [`wasm_bindgen`](https://rustwasm.github.io/wasm-bindgen/) is a bridge between Rust and JavaScript. These packages aren't usable for AIR though or are entirely incompatible, like `chrono` (a date-time library). Other packages don't make sense, like `tokio`, a multi-threaded asynchronous runtime for Rust, since AIR is already asynchronous, and also, some parts of `tokio` do use parts of `std` related to time (like `std::time::Instant::now()`).
- To easily bridge between the AIR API and Rust, a macro similiar to the ones that `wasm_bindgen` has can be defined.

**Reply by hydroper** [reply]
Here are the `std` things I was referring to: https://github.com/rust-lang/rust/blob/9622cdee1ad9657dd93d67a9ffdbc16e196ee4fa/library/std/src/sys/wasm/mod.rs#L48
